File: src/engine/adapters/sqlite/console.ts
import type { StatementSync } from 'node:sqlite';
import {
  type ColumnDescriptor,
  createTabularPageBuilder,
  type PagePosition,
  type TabularPage,
} from '../../../shared/protocol/page';
import type { OpCtx } from '../adapter';
import { AdapterError } from '../errors';
import { singleStatusPage } from '../sql-text';
import type { SqliteHandle } from './client';
import { mapError } from './errors';
import { checkNotCancelled, prepareOne } from './query';
import { toCellText, typeClassFor } from './read';

// F5: `StatementSync.columns()` reports each column's *declared* origin type ('INTEGER', 'TEXT',
// null for an expression or an untyped column) — the same vocabulary `pragma_table_xinfo` uses,
// unlike MariaDB's own wire-type vocabulary (mysql-family/console.ts). `typeClassFor` is exactly
// the read path's function, reused unchanged.
function columnsFor(stmt: StatementSync): ColumnDescriptor[] {
  return stmt.columns().map((c) => ({
    name: c.name,
    dataType: c.type ?? '',
    typeClass: typeClassFor(c.type),
    // execute() never consults the catalog — nullability/PK-ness are unknowable here; console
    // results are always read-only regardless (mirrors mysql-family/console.ts's own note).
    nullable: true,
    isPrimaryKey: false,
    generated: false,
  }));
}

// F5: `columns().length === 0` is exactly how a non-row-returning statement (INSERT/UPDATE/DELETE/
// DDL/pragma) is told apart from a SELECT — verified against node:sqlite directly, not assumed.
function runOneStatement(stmt: StatementSync): TabularPage {
  const descriptors = columnsFor(stmt);
  try {
    if (descriptors.length === 0) {
      const result = stmt.run();
      return singleStatusPage(`${result.changes} row(s) affected`, 'text');
    }

    // D5: `.iterate()`, not `.all()` — a user's own console statement is unbounded, unlike the
    // read path's own `LIMIT pageSize + 1`, and streaming avoids holding a second full copy of
    // the result alongside the page builder's own buffers.
    stmt.setReturnArrays(true);
    const builder = createTabularPageBuilder(descriptors);
    let rowCount = 0;
    for (const row of stmt.iterate() as Iterable<unknown[]>) {
      builder.appendRow(row.map(toCellText));
      rowCount++;
    }
    const position: PagePosition = {
      offset: 0,
      pageSize: rowCount,
      hasMore: false,
      nextToken: null,
      prevToken: null,
      strategy: 'offset',
    };
    return builder.finish(position);
  } catch (err) {
    throw mapError(err);
  }
}

export function execute(h: SqliteHandle, ctx: OpCtx, statements: string[]): TabularPage[] {
  if (statements.length === 0) throw new AdapterError('E_QUERY', 'no statements to execute');
  // One op-log row for the whole batch (P5 D9's precedent), not one per statement.
  ctx.setCommand(statements.join(';\n'));

  const pages: TabularPage[] = [];
  for (const sql of statements) {
    checkNotCancelled(ctx);
    // D3: a user's own console statement can touch real table data just as easily as the read
    // path can.
    const stmt = prepareOne(h, sql, { readBigInts: true });
    pages.push(runOneStatement(stmt));
  }
  return pages;
}

File: apps/kira-studio/internal/adapters/sqlite/console.go
package sqlite

import (
	"context"
	"database/sql"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/page"
)

// columnsFor is console.ts's own — F5: a column's *declared* origin type ('INTEGER', 'TEXT', "" for
// an expression or an untyped column), the same vocabulary pragma_table_xinfo uses. typeClassFor is
// exactly the read path's function, reused unchanged. execute() never consults the catalog —
// nullability/PK-ness are unknowable here; console results are always read-only regardless.
func columnsFor(names, declTypes []string) []page.ColumnDescriptor {
	columns := make([]page.ColumnDescriptor, len(names))
	for i, name := range names {
		columns[i] = page.ColumnDescriptor{
			Name: name, DataType: declTypes[i], TypeClass: typeClassFor(declTypes[i]),
			Nullable: true, IsPrimaryKey: false, Generated: false,
		}
	}
	return columns
}

// runOneStatement is console.ts's own — F5: a non-row-returning statement (INSERT/UPDATE/DELETE/
// DDL/pragma) is told apart from a SELECT by QueryContext's own zero-column result, the same signal
// mysqlfamily's console.go uses (MY-1 confirmed it there; verified independently for
// modernc.org/sqlite too, not assumed from that unrelated driver's behaviour).
func runOneStatement(ctx context.Context, conn *sql.Conn, sqlText string) (page.TabularPage, error) {
	if err := assertSingleStatement(sqlText); err != nil {
		return page.TabularPage{}, err
	}
	rows, err := conn.QueryContext(ctx, sqlText)
	if err != nil {
		return page.TabularPage{}, mapError(err)
	}
	defer rows.Close()

	types, err := rows.ColumnTypes()
	if err != nil {
		return page.TabularPage{}, mapError(err)
	}
	if len(types) == 0 {
		// A non-row-returning statement is read through QueryContext (never ExecContext) so the
		// zero-column signal above is what tells the two apart — but database/sql's *sql.Rows
		// exposes no RowsAffected count at all (that lives only on the driver.Result ExecContext
		// returns), so unlike node:sqlite's own `${result.changes} row(s) affected`, this reports a
		// generic "OK" — the same capability loss B4 already documented for mysql-family, for the
		// same reason: this is a database/sql API shape, not a driver-specific gap.
		return adapters.SingleStatusPage("OK", "text"), nil
	}

	names := make([]string, len(types))
	declTypes := make([]string, len(types))
	for i, t := range types {
		names[i] = t.Name()
		declTypes[i] = t.DatabaseTypeName()
	}
	columns := columnsFor(names, declTypes)

	builder := page.NewTabularPageBuilder(columns)
	rowCount := 0
	for rows.Next() {
		vals := make([]any, len(types))
		dest := make([]any, len(types))
		for i := range vals {
			dest[i] = &vals[i]
		}
		if err := rows.Scan(dest...); err != nil {
			return page.TabularPage{}, mapError(err)
		}
		cells := make([]*string, len(vals))
		for i, v := range vals {
			cells[i] = toCellText(v)
		}
		if err := builder.AppendRow(cells); err != nil {
			return page.TabularPage{}, err
		}
		rowCount++
	}
	if err := rows.Err(); err != nil {
		return page.TabularPage{}, mapError(err)
	}
	return builder.Finish(page.UnpagedPosition(rowCount)), nil
}

// execute is console.ts's own execute.
func execute(ctx context.Context, conn *sql.Conn, op *adapters.OpCtx, statements []string) ([]page.Page, error) {
	if len(statements) == 0 {
		return nil, adapters.New(adapters.CodeQuery, "no statements to execute", nil)
	}
	// One op-log row for the whole batch (P5 D9's precedent), not one per statement.
	op.SetCommand(joinSemicolons(statements))

	pages := make([]page.Page, len(statements))
	for i, stmt := range statements {
		if err := adapters.CheckCancelled(ctx); err != nil {
			return nil, err
		}
		p, err := runOneStatement(ctx, conn, stmt)
		if err != nil {
			return nil, err
		}
		pages[i] = p
	}
	return pages, nil
}

func joinSemicolons(parts []string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += ";\n"
		}
		out += p
	}
	return out
}
